src/Util/Security.php

```diff
@@ -0,0 +1,273 @@
+	 * The security token has to base on at least one value that an attacker can't know - here it's the session ID and the private key.
+	 * In this implementation, a security token is reusable (if the user submits a form, goes back and resubmits the form, maybe with small changes;
+	 * or if the security token is used for ajax-calls that happen several times), but only valid for a certain amout of time (3hours).
+	 * The "typename" seperates the security tokens of different types of forms. This could be relevant in the following case:
+	 *    A security token is used to protekt a link from CSRF (e.g. the "delete this profile"-link).
+	 *    If the new page contains by any chance external elements, then the used security token is exposed by the referrer.
+	 *    Actually, important actions should not be triggered by Links / GET-Requests at all, but somethimes they still are,
+	 *    so this mechanism brings in some damage control (the attacker would be able to forge a request to a form of this type, but not to forms of other types).
+	 */
+	public static function get_form_security_token($typename = '')
+	{
+		$a = get_app();
+
+		$timestamp = time();
+		$sec_hash = hash('whirlpool', $a->user['guid'] . $a->user['prvkey'] . session_id() . $timestamp . $typename);
+
+		return $timestamp . '.' . $sec_hash;
+	}
+
+	public static function check_form_security_token($typename = '', $formname = 'form_security_token')
+	{
+		$hash = null;
+
+		if (!empty($_REQUEST[$formname])) {
+			/// @TODO Careful, not secured!
+			$hash = $_REQUEST[$formname];
+		}
+
+		if (!empty($_SERVER['HTTP_X_CSRF_TOKEN'])) {
+			/// @TODO Careful, not secured!
+			$hash = $_SERVER['HTTP_X_CSRF_TOKEN'];
+		}
+
+		if (empty($hash)) {
+			return false;
+		}
+
+		$max_livetime = 10800; // 3 hours
+
+		$a = get_app();
+
+		$x = explode('.', $hash);
+		if (time() > (IntVal($x[0]) + $max_livetime)) {
+			return false;
+		}
+
+		$sec_hash = hash('whirlpool', $a->user['guid'] . $a->user['prvkey'] . session_id() . $x[0] . $typename);
+
+		return ($sec_hash == $x[1]);
+	}
+	
+	private static function check_form_security_std_err_msg()
+	{
+		return L10n::t("The form security token was not correct. This probably happened because the form has been opened for too long \x28>3 hours\x29 before submitting it.") . EOL;
+	}
+
+	public static function check_form_security_token_redirectOnErr($err_redirect, $typename = '', $formname = 'form_security_token')
+	{
+		if (!check_form_security_token($typename, $formname)) {
+			$a = get_app();
+			logger('check_form_security_token failed: user ' . $a->user['guid'] . ' - form element ' . $typename);
+			logger('check_form_security_token failed: _REQUEST data: ' . print_r($_REQUEST, true), LOGGER_DATA);
+			notice(check_form_security_std_err_msg());
+			goaway(System::baseUrl() . $err_redirect);
+		}
+	}
+
+	public static function check_form_security_token_ForbiddenOnErr($typename = '', $formname = 'form_security_token')
```

Comment:
Since this method sends headers, please move this method to `BaseModule::checkFormSecurityTokenForbiddenOnError`.